tests/syntax/test_specifiers.py

```diff
@@ -263,50 +615,227 @@ def test_in_heading_distribution():
+        assert pos.y == pytest.approx(-3, abs=2*scene.egoObject.contactTolerance) or \
+               pos.y == pytest.approx(3, abs=2*scene.egoObject.contactTolerance)
+
+def test_on_modifying_surface_onDirection():
+    scenario = compileScenic("""
+        workspace = Workspace(BoxRegion(dimensions=(10,10,10)))
+        box = BoxRegion(dimensions=(5,5,5), onDirection=(0,1,0)).getSurfaceRegion()
+        ego = new Object in workspace, on box, with shape ConeShape(),
+            with onDirection (0,1,0)
+    """)
+    for i in range(30):
+        scene = sampleScene(scenario, maxIterations=1000)
+        pos = scene.egoObject.position
+        assert pos.y == pytest.approx(-3, abs=2*scene.egoObject.contactTolerance) or \
+               pos.y == pytest.approx(3, abs=2*scene.egoObject.contactTolerance)
+
+def test_on_mistyped():
+    with pytest.raises(TypeError):
+        compileScenic('ego = new Object on "foo"')
+
+def test_on_incompatible():
+    with pytest.raises(SpecifierError):
+        compileScenic("""
+            box = BoxRegion()
+            ego = new Object on box, on box
+        """)
+# Following
 def test_following():
     ego = sampleEgoFrom("""
         vf = VectorField("Foo", lambda pos: 90 deg * (pos.x + pos.y - 1),
                          minSteps=4, defaultStepSize=1)
-        ego = Object at 1@1
-        ego = Object following vf for 4
+        ego = new Object at 1@1
+        ego = new Object following vf for 4
     """)
-    assert tuple(ego.position) == pytest.approx((-1, 3))
+    assert tuple(ego.position) == pytest.approx((-1, 3, 0))
     assert ego.heading == pytest.approx(math.radians(90))
 
 def test_following_from():
     ego = sampleEgoFrom("""
         vf = VectorField("Foo", lambda pos: 90 deg * (pos.x + pos.y - 1),
                          minSteps=4, defaultStepSize=1)
-        ego = Object following vf from 1@1 for 4
+        ego = new Object following vf from 1@1 for 4
     """)
-    assert tuple(ego.position) == pytest.approx((-1, 3))
+    assert tuple(ego.position) == pytest.approx((-1, 3, 0))
     assert ego.heading == pytest.approx(math.radians(90))
 
 def test_following_random():
     ego = sampleEgoFrom("""
         vf = VectorField('Foo', lambda pos: -90 deg)
         x = Range(1, 2)
-        ego = Object following vf from 1@2 for x, facing x
+        ego = new Object following vf from 1@2 for x, facing x
     """)
-    assert tuple(ego.position) == pytest.approx((1+ego.heading, 2))
+    assert tuple(ego.position) == pytest.approx((1+ego.heading, 2, 0))
 
-## Heading specifiers
+## Orientation Specifiers ##
 
+# Facing
+def test_facing_heading():
+    ego = sampleEgoFrom('ego = new Object facing 90 deg')
+    assert ego.heading == pytest.approx(math.radians(90))
+
+def test_facing_orientation():
+    ego = sampleEgoFrom('ego = new Object facing (90 deg, 90 deg, 90 deg)')
+    assert ego.orientation.approxEq(Orientation.fromEuler(math.pi/2, math.pi/2, math.pi/2))
+
```

Comment:
We should add versions of these tests that check the final orientation is correct even when `parentOrientation` is not the default.